### PR TITLE
PLT-7098 Adapt Determinism proof to SOPs

### PR DESCRIPTION
**Pull request by mjaskelioff**
This PR extends the determinism proof of the reduction semantics to handle SOPs.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

**Review by kwxm — APPROVED** [review]
Heroic work!  I didn't understand the details of the proofs, but I was usually able to work out roughly what the types/theorems are saying and since Agda accepts it all I'm willing to believe it.

One thing that might have been helpful might have been to have brief comments explaining some of the naming conventions.  For example, `TSS` appears quite a lot and I wasn't able to work out what it stands for, and similarly for `ASS`. Another example is `SOP`:  I know  what that means but a random reader might not. One-line comments saying what these things stand for would be a big help.  However, I don't know where would be a good place to put such comments since the notation's spread throughout the codebase and it might not be easy to find the relevant comment when you come across a name you don't understand.  Maybe I need a better way of browsing Agda code than Emacs and grep.

File: plutus-metatheory/src/Algorithmic/ReductionEC/Determinism.lagda.md
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply plutus-metatheory/src/Algorithmic/ReductionEC/Determinism.lagda.md:1]: Yes. I will add more comments.

**Comment by mjaskelioff** [comment]
> Heroic work! I didn't understand the details of the proofs, but I was usually able to work out roughly what the types/theorems are saying and since Agda accepts it all I'm willing to believe it.
> 
Thank you for taking a look at it.

> One thing that might have been helpful might have been to have brief comments explaining some of the naming conventions. For example, `TSS` appears quite a lot and I wasn't able to work out what it stands for, and similarly for `ASS`. 

With respect to variable naming, I'm just following the convention to name `xs` a list of `x`s and `xss` a list of list of `x`s (that is a list of `xs`). In this case is really a vector of lists, rather than a list of lists, but I think the principle is the same.

> Another example is `SOP`: I know what that means but a random reader might not. One-line comments saying what these things stand for would be a big help. However, I don't know where would be a good place to put such comments since the notation's spread throughout the codebase and it might not be easy to find the relevant comment when you come across a name you don't understand. Maybe I need a better way of browsing Agda code than Emacs and grep.

When you load the file in Agda, you can follow a name to its definition, so I guess the place to put any explanation is right next to the definition.

**Comment by kwxm** [comment]
> With respect to variable naming, I'm just following the convention to name `xs` a list of `x`s and `xss` a list of list of `x`s (that is a list of `xs`). In this case is really a vector of lists, rather than a list of lists, but I think the principle is the same.

Ah, I see.  I'm not used to seeing that in upper case!  I assumed that `TSS` was an abbreviation for something.

File: plutus-metatheory/src/Algorithmic/ReductionEC/Determinism.lagda.md [hunk 1]
```diff
@@ -191,6 +222,13 @@ proj· : ∀{A A' B}{t : ∅ ⊢ A ⇒ B}{t' : ∅ ⊢ A' ⇒ B}{u : ∅ ⊢ A}{
       × substEq (∅ ⊢_) p u ≡ u'
 proj· refl = refl ,, refl ,, refl
 
+subst-case : ∀ {B} {A} {B'} {n}
+               {TSS = tss : Vec.Vec (List (∅ ⊢Nf⋆ *)) n}
```
[the remark this answers is not in the document]
Comment (mjaskelioff) [reply plutus-metatheory/src/Algorithmic/ReductionEC/Determinism.lagda.md]: Renamed all the vector of lists to `Tss`, and tried to be more uniform in the choice of variable names for lists and lists of lists.

**Pull request merged**